### ISSUE TITLE
Access Outputs from CF template in serverless.yml 

### ISSUE THREAD

**Issue by picsoung** [serverless/serverless#2240]
Hi y'all,

Trying to transition to v1, and porting my CF template too.

I am wondering if I am facing the same issue as in #1699. Trying to access to `outputs` under `resources` from Cloudformation.

In my CF template, I am creating a VPC, subnets and such that I want to reuse in the config of my functions.

my `serverless.yml`

``` yml
provider:
  name: aws
  vpc:
    securityGroupIds:
     - Ref: vpcSecurityGroup
    subnetIds:
      - Ref: PrivateSubnet1
      - Ref: PrivateSubnet2
```

then under resources

``` yml
resources:
  Resources:
    VPC:
      Type: AWS::EC2::VPC
      Properties:
        CidrBlock: 10.0.0.0/16
    PrivateSubnet1:
      Type: AWS::EC2::Subnet
      Properties:
        VpcId:
          Ref: VPC
        CidrBlock: 10.0.1.0/24
    SecurityGroup:
      Type: AWS::EC2::SecurityGroup
      Properties:
        GroupDescription: Limits security group egress traffic
        VpcId:
          Ref: VPC
        SecurityGroupIngress:
        - IpProtocol: "-1"
        SecurityGroupEgress:
        - IpProtocol: "-1"
          CidrIp: 0.0.0.0/0
Outputs:
    vpcSecurityGroup:
      Description: Id of the SecurityGroup
      Value:
        Fn::GetAtt:
          - SecurityGroup
          - GroupId
```

But this setup gives the following error:

```
Template format error: Unresolved resource dependencies
     [vpcSecurityGroup] in the Resources block of the template
```

I also tried to access directly, using `Fn:GetAtt` without success :(

Would you confirm you can not access `Outputs` values, yet?
Thanks

**Comment by s0enke** [comment]
Yes, I can confirm that you cannot access the CFN outputs currently. Maybe it could help to create the VPN/Subnet resources outside of serverless (e.g. with plain CloudFormation) and then pass e.g. `PrivateSubnet1` to serverless as a parameter or a env var.

**Comment by WooDzu** [comment]
A new CloudFormation feature released recently enables exporting stack outputs which can be referenced from other templates. http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/walkthrough-crossstackref.html

**Comment by andymac4182** [comment]
This is because outputs are for outputting to outside of the current CloudFormation. You want to reference `SecurityGroup` directly. 

This isn't a serverless error but a CloudFormation one.

**Comment by rowanu** [comment]
How would this not result in circular references? When you first launch the service the outputs do not exist, so how can the `provider` object reference them?

Also, please follow the suggested feature/bug templates in the future.

**Comment by andymac4182** [comment]
If you allow CloudFormation refs as part of the VPC config then it will be handled in CloudFormation and not be a circular reference. Well that is purely how I think it would work after 2min of thinking about it.

**Comment by rowanu** [comment]
If you're just using Refs to CFN Logical Ids, then it might work (but would be hard to validate, and would break your stack if you got it wrong).

You cannot reference CFN Outputs (as defined in the `resources` example above) before the stack has been created.

**Comment by picsoung** [comment]
In Serverless 0.5, it was handled separately with `s-resources-cf.json` at the root level, and `s-function.json` at the function level. And you can reference to outputs from CF.
I am trying to have a similar behavior.

Or something where I will first deploy the stack, then my functions like
`project init` and `deploy` in the past.

Having a separate file, would work, but it defeats the idea of having everything handled by serverless :(

**Comment by AndrewFarley** [comment]
+1 - Really need this feature.

Without this feature, basically we all have to run a separate "init environment" script (terraform for me) which outputs things to a .json file such as SQS URL or DynamoDB ARN which I then refer to in the serverless.yml file.  It would be nice/ideal to not have to _also_ run CF/TF separately from Serverless to provision and/or version an environment since CF is now supported in serverless v1.0+.  I know the output feature of CF is brand new... but this feature feels like it absolutely should already be in Serverless.  I've recently migrated a few of my own CF scripts to use this cross-linking.

In order to make this possible, your Serverless CF stack may need to consist of more than one stack, having one stack able to initialize and then have the second stack refer to the outputs of previous stack. eg: [Cross-Stack CF](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/walkthrough-crossstackref.html).  

This would allow for two stacks, the first, lower-level stack being the environment provisioning (Eg: creating DynamoDB tables, SQS queues, SNS topics, etc) and then the next/final-level stack being the actual serverless lambda/api-gateway/cloudwatch-events code deployment.  I envision re-deployments of a serverless codebase to only touch the second stack, unless you customize the evnironment CF in serverless.yml then it also re-runs the first stack.

**Comment by erikerikson** [comment]
RE: cross-stack references...

Maybe I misunderstand what people are asking for here but I have been using the following in my services using a published version of SLS:

```
service: my-thing-service
# ...
resources:
  #...
  Outputs:
    MyThingArn:
      Description: The ARN for My Thing
      Value:
        'Fn::GetAtt': [ MyThing, Arn ]
      Export:
        Name: '${self:service}:${self:custom.stage}:MyThingArn'
```

Then, in another service, I am using the following:

```
functions:
  userOfThing:
    handler: userOfThing.handler
    environment:
      MY_THING:
        'Fn::ImportValue': 'my-thing-service:${self:custom.stage}:MyThingArn'
```

And this totally allows me to have no need for the "requested feature" here.  There are a few things to keep in mind (see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-importvalue.html) but accepting those limitations and considerations, I think everything you'd want to do can be done.

For @picsoung's original question, which appears to be a "within stack" use case, I don't know why the original logical ID wouldn't be used:

```
provider:
  name: aws
  vpc:
    securityGroupIds:
     - 'Fn::GetAtt': SecurityGroup
    subnetIds:
      - Ref: PrivateSubnet1
      - Ref: PrivateSubnet2
```

Please, if I am missing something, let me know but I think this is "old way of doing things" mixing with the new way where you don't need to do that thing anymore.

**Comment by rowanu** [comment]
@erikerikson nope, you're right - this issue pre-dates the cross-stack references announcement.

Any other outputs should be retrieved by using API calls (i.e. via an SDK).

**Comment by erikerikson** [comment]
Ah, cool.  Hopefully it'll help someone searching issues for a solution then.  :)

**Comment by kennu** [comment]
@erikerikson's advice to use Ref's in environment variables is excellent. I couldn't figure it out from serverless.com docs, so maybe it could be mentioned in the Variables and Resources sections.

**Comment by laardee** [comment]
@picsoung try to remove `-`
```
    ELASTICACHE_ENDPOINT:
       Fn::GetAtt: elasticCache.RedisEndpoint.Address
     SNS_TOPIC_ARN:
       Ref: SNStopic
```

Like the FEED_URL in this https://github.com/SC5/serverless-blog-workshop/blob/rss/serverless.yml

**Comment by bjfletcher** [comment]
I'm trying to use `Fn::ImportValue` but it doesn't work.

Template:

```
resources:
  Resources:
    BFCDN:
      Type: AWS::CloudFront::Distribution
      Properties:
        DistributionConfig:
          Origins:
            - DomainName:
              "Fn::Sub":
                - "${bucket}.s3.amazonaws.com"
                -
                  - bucket:
                    "Fn::ImportValue": BFBucket
```

Error when running `deploy`:

```
  Serverless Error ---------------------------------------

     Invalid variable reference syntax for variable bucket.
     You can only reference env vars, options, & files. You
     can check our docs for more info.
```

I see there's an example by @erikerikson in https://github.com/serverless/serverless/issues/2240#issuecomment-268411199 how to use this with functions, but that's because environment is passed straight in (as far as I can tell)... and doesn't explain how to import values for use within the `serverless.yml` template.

**Comment by erikerikson** [comment]
@bjfletcher "Fn::ImportValue" is resolved by CloudFormation so for something that configures the local behavior of the framework (such as the bucket to deploy to) it won't work.  You'll need to hard code the bucket name, either in a config file that you load in your serverless.yml or directly in serverless.yml

**Comment by erikerikson** [comment]
There might be a feasible way to do that but no one has offered an implementation yet :D

**Comment by picsoung** [comment]
thanks @laardee 
indeed it worked like this:

```
 ELASTICACHE_ENDPOINT:
     Fn::GetAtt:
       - elasticCache
       - RedisEndpoint.Address
   SNS_TOPIC_ARN:
     Ref: SNStopic
```

But I could not take advantage of the `Outputs` section of my CF template.

[Here](https://github.com/haw-itn/serverless-web-monitor/blob/b20579e132cf4f9445b1529499b4f9121b323ee4/serverless.yml)
I've seen someone using `"${self:resources.Outputs.ResourceName.ResourceAttribute}"`

That could be another way, and it worth documenting it

**Comment by ccverak** [comment]
Hi @picsoung, @laardee can you please help me with an example on how to setup the Outputs/References DynamoDB streams and a function using it? I was not able to replicate the example above

Not working example:
```
functions:
  myFunction:
    handler: handler.myFunction
    events:
      - stream:
          arn: ${self:resources.Outputs.MyTableStreamArn.Value}
          startingPosition: 'LATEST'
          batchSize: 100

Outputs:
    MyTableStreamArn:
      Description: The ARN of the DynamoDB stream of myTable
      Value:
        Fn::GetAtt: [ myTable, StreamArn ]
```

**Comment by pmuens** [comment]
This should be possible with the new `cf` variable type where one can reference stack outputs.

**Comment by phil-hachey** [comment]
@pmuens I tried using `cf:` in my provider.environment block. It works great when my stack is already created, but fails on a new stack creations with:
```
  Serverless Error ---------------------------------------
 
     Stack with id mystack does not exist
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Forums:        forum.serverless.com
     Chat:          gitter.im/serverless/serverless
 
  Your Environment Information -----------------------------
     OS:                 linux
     Node Version:       5.12.0
     Serverless Version: 1.13.1
```

Am I missing something? I've also tried the approach of using `${self:resources.Outputs.ResourceName.ResourceAttribute}`, which works when deployed on Lambda, but not locally.

**Comment by pmuens** [comment]
@phil-hachey the `cf:` variable expects the stack you're referencing to already be deployed.

Serverless will issue a SDK call which queries the stack behind the scenes.

**Comment by phil-hachey** [comment]
@pmuens right, so if I wanted to use stack outputs as environment variables, what would be the best approach?

**Comment by pmuens** [comment]
@phil-hachey you could use the `cf:` Serverless Variable prefix and add them to the `Environment` config of your function like this:

```yml
functions:
  hello:
    handler: handler.hello
      environment:
        SOMEVAR: ${cf:another-service-dev.outputvalue}
```

You could also use [`Fn::ImportValue`](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-importvalue.html) in the `resources` section and add the values by merging them into the generated Function template (but that's a little bit more cumbersome).

Just remember that the generated CloudFormation template will contain the variables in plain text.

**Comment by phil-hachey** [comment]
@pmuens Sorry I meant use the stack outputs of my **current** stack. Here's my scenario:

I have an AWS Step Function state machine whose ARN I would like to reference in one of my handlers. Since CloudFormation doesn't support Name for the StateMachine I can't predict the generated name in my handler. To get around this I would have liked to use `cf:` to get the ARN out of my stack outputs and set it as an environment variables. For example:
```
environment:
  StateMachineArn: ${cf:${self:custom.stack_name}.StateMachineArn}
  ActivityArn: ${cf:${self:custom.stack_name}.ActivityArn}
```

The problem is if I'm trying to create a new stack, it tries to query `custom.stack_name` at the very beginning of `serverless deploy` and fails because the current stack doesn't exist (obviously). I'm not sure if I'm approaching this wrong? Should I be creating dependency resources in a different stack?

**Comment by HyperBrain** [comment]
Why don't you just use:
```
environment:
  StateMachineArn:
    Fn::GetAtt
      - ...
      - ...
  ...

<or>

  StateMachineArn:
    Ref: xxxxxx
```

This should be evaluated on CF deployment. Use the same GetAtt or Ref here, that you use to set your Output variable.

**Comment by phil-hachey** [comment]
@HyperBrain I could, but then I run into the same problem as when I try to do `${self:resources.Outputs.ResourceName.ResourceAttribute}` where I it doesn't work with `serverless invoke local`. The ARN value just evaluates to `[object Object]`

**Comment by HyperBrain** [comment]
The solution for the local invoke is the export-env plugin ;-) by @arabold
https://github.com/arabold/serverless-export-env

**Comment by pmuens** [comment]
> The problem is if I'm trying to create a new stack, it tries to query custom.stack_name at the very beginning of serverless deploy and fails because the current stack doesn't exist (obviously).

Yes, the `cf:` Serverless Variable prefix issues a SDK request before compiling the template. There is no workaround for now since the main purpose of this variable is to enable cross-service communication between different stacks.

Have you had time to look at the plugin @HyperBrain posted? 🔝

**Comment by phil-hachey** [comment]
@pmuens @HyperBrain Yep I looked at all your suggestions. Thanks to both of you for all your help! I forgot to mention though that I'm developing in Python using the serverless-wsgi plugin so I wrote my own plugin that basically does the same as serverless-export-env:
https://github.com/phil-hachey/serverless-wsgi-export-env

I tried integrating serverless-export-env, but I ran into an issue where my stack had so many resources that some of them were dropped from the AWS CF [DescribeStackResources](http://docs.aws.amazon.com/cli/latest/reference/cloudformation/describe-stack-resources.html) call. This is a bug with the AWS API since it doesn't look like it implements any paging.

**Comment by pmuens** [comment]
Thanks for getting back @phil-hachey 👍

The plugin looks good. Thanks for sharing. Hmm, yes it seems like `ListStackResources` returns more results 🤔 (AFAIK we've switched back and forth between those two and ListStackResources were missing some critical information we needed. Not sure what this exactly was though.).

**Comment by tommedema** [comment]
Is there any documentation on how to create a serverless.yml that results in 100% of your infrastructure being version controlled, i.e. in code / no manual clicking around in GUIs?

I've seen mentions of `Ref:`, `cf:`, `Fn::GetAtt:` and `${self:resources.Outputs...`, all somewhat confusing. I couldn't find a clear documentation on their purpose and differences.

**Comment by AndrewFarley** [comment]
@tommedema I haven't published a blog about this (I should though) but what you're looking for is combining Terraform with Serverless.  Terraform can be used to build your more static infrastructural elements that might be shared with multiple serverless stacks (such as a VPC, RDS, ElastiCache, etc), and then it saves that information into a state file (remote or local) which you can parse and refer to (since it's just .json) inside your serverless stacks.  If you "wrap" your serverless deploy script (aka, make a deploy.sh script that calls serverless deploy) in that deploy script it can first pull the remote state from terraform, so that you can pull variables, (like the RDS hostname, or the VPC id, or whatever) to properly deploy and have a functional serverless stack.

**Comment by erikerikson** [comment]
@tommedema What you've listed is a combination of CloudFormation intrinsic functions (resolved remotely by that service) and Serverless Framework variables (resolved locally by the framework).  If the Terraform approach isn't what you were looking for (I haven't used it myself and can't advise) you can define a CloudFormation resource in a `resources.Resources` section of your serverless.yml (the`Resources`bit corresponds to the standard CloudFormation template clause).

See the following, respectively:
http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html
https://serverless.com/framework/docs/providers/aws/guide/variables/

**Comment by AndrewFarley** [comment]
@erikerikson Raises a good point... you don't need Terraform, you can have your complete infrastructure managed by a serverless stack... but you might not find any serverless examples that show you this.  Your best bet is to google around for demo CloudFormation Stacks that implement what you want (such as a VPC, RDS, etc).  You then do what Eric mentioned, and put them into your Resources area.

However... having said that... that is not best practice or advice I would ever give a client.  You should do similar to what I mentioned above, and build your "stack" in CloudFormation (or Terraform) which manages your infrastructure, and limit the scope of what your serverless stack does.  This makes deployments MUCH quicker, and much less error-prone.

A great example I can give, is if you upgraded serverless and you ran a stack update, it would (in the past) destroy your API gateway (because eg the naming scheme changed) and create a new one, which would give you a new URL.  And if you had this URL programmed into, say, an iPhone app, then suddenly your service has downtime (had this happen to a client).  The same thing could happen if you made a slight change to your cloudformation which changed, for example, a parameter in your mysql server that required a re-creation of the mysql server.  CF would destroy and re-create that server, and since serverless doesn't give you the ability (yet) to preview the CF changes it will apply, you could easily accidentally destroy part of your infrastructure upon deploying.  This feature is WIP, implementing Change Sets or "diffs" of sorts to detect the changes made on the back end.

In summary: You probably should not do what you want to do, manage your entire infra + serverless in a single stack.  Split the stacks, you can refer to outputs from one stack in another stack.  There's a plugin to help you with that that I've used:

https://www.npmjs.com/package/serverless-plugin-additional-stacks

And if you do a lot of CloudFormation in your serverless you might like a helper to help put those intrinsic functions directly in your serverless stack without having to escape it weird...

https://www.npmjs.com/package/serverless-pseudo-parameters

**Comment by tommedema** [comment]
With all of this information in mind I will try to set something up from scratch, preferably without Terraform (to keep things simple). However could someone please refer me to the documentation for `Ref:`? I've seen it used in serverless.yml files but have not seen it's usage being described.

**Comment by AndrewFarley** [comment]
@tommedema Erik linked you this above. This is a CloudFormation feature, not a Serverless feature.  Serverless supports embedding cloudformation, and any features of CF are available to us.  Anyways, the link is...

See: http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html

**Comment by tommedema** [comment]
@AndrewFarley I get it now, thanks.

Since `Ref:` is supposed to reference to cloudformation outputs, can you confirm if it is possible to refer to cloudformation resources that have not yet been created (manually)? I.e. a `serverless deploy` will first setup all `Resources` using cloudformation, prior to referencing to these resources in the other `serverless.yml` sections? As @pmuens [mentioned](https://github.com/serverless/serverless/issues/2240#issuecomment-304793165) this is not (yet) possible for `cf:` syntax (or has this been fixed?).

**Comment by AndrewFarley** [comment]
@tommedema Absolutely, ref can refer to not-yet-created items.  This creates a dependency tree in CloudFormation (similarly in TerraForm) which requires that the referred-to item gets created first, and then its output can be used in whatever is referencing it.  For example, when you want to launch an EC2 Instance, you Ref: to the VPC you want to put it in.  CloudFormation/TF will then figure out that it must create the VPC first, then feed the VPC ID (via the Ref:) to the EC2 Instance stanza to be able to successfully create it.
